### PR TITLE
[WIP] only clients that support cookies should get a browser session

### LINKED ISSUE #24730: if DAV clients log in via username and token, the session is killed
### Steps to reproduce
1. Generate a token
2. Configure your sync client with user:token
3. Start syncing

### Expected behaviour
The sync clients can authenticate successfully and a browser session is created (#24543 will fix that)

### Actual behaviour
The sync client can authenticate successfully, but as soon as the token is validated (can happen on login and on subsequent requests) the session is killed as the login password is wrong (when creating the session token the login password is actually a token).

The problematic code is here https://github.com/owncloud/core/blob/65eae43ab122bf1d04a5d6f8bee8e8b5ec439276/apps/dav/lib/Connector/Sabre/Auth.php#L106

A session token is created on successful client login. However, if the client uses a token instead of a password, we'd need to restore (decrypt) the login password first and then create the session token.

cc @LukasReschke

### PR DESCRIPTION
**Pull request by ChristophWurst**
POC WIP to fix #24730 and maybe #24543
- check if client supports cookies, if yes create a session token
- do not log in when creating a session token (regernates the session id) and log in explicitly instead
- restore login password if dav authenticates with token as password (#24730)

@LukasReschke @rullzer @PVince81 thoughts/objections?

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/Auth.php [hunk 1]
```diff
@@ -104,7 +104,12 @@ protected function validateUserPass($username, $password) {
 		} else {
 			\OC_Util::setUpFS(); //login hooks may need early access to the filesystem
 			if($this->userSession->login($username, $password)) {
-				$this->userSession->createSessionToken($this->request, $username, $password);
+				if (!is_null($this->request->getCookie('supports_cookies'))) {
+					unset($_COOKIE['supports_cookies']);
```
Comment (ChristophWurst) [inline apps/dav/lib/Connector/Sabre/Auth.php]: do we have something mockable for this?

File: apps/dav/lib/Connector/Sabre/Auth.php [hunk 2]
```diff
@@ -104,7 +104,12 @@ protected function validateUserPass($username, $password) {
...
+					$this->userSession->createSessionToken($this->request, $username, $password);
+				} else {
+					setcookie('supports_cookies', true, time() + 3600);
```
Comment (ChristophWurst) [inline apps/dav/lib/Connector/Sabre/Auth.php]: same here, not mockable

**Comment by guruz** [comment]
I don't really like it because I think if we have a 1:1 mapping of token and session anyway (still to be discussed if we do) then we can just deduce the session from the token.

**Comment by ChristophWurst** [comment]
@guruz not sure if I understand you. Which token? Session token or device token? How could we reuse a PHP session if the client does not send the session cookie? I don't get it

**Comment by guruz** [comment]
Like this:
```
<ChristophWurst> guruz: how would that work? how can I "use the same session" if the client does not send a cookie?
<ChristophWurst> I guess you're talking about php sessions
<guruz> ChristophWurst: he sends the same token via Auth header. Then you use session_name(something_based_on(authHeaderToken)); session_start();
```
Or maybe I don't understand this token stuff completely yet?

**Comment by ChristophWurst** [comment]
Hm. Looks like [session_name()](http://php.net/manual/en/function.session-name.php) only sets the name which is stored in the session cookie. That would mean you still have two separate sessions, all they share is their name. Hence session name !== session id.

> The session name references the name of the session, which is used in cookies and URLs (e.g. PHPSESSID). It should contain only alphanumeric characters; it should be short and descriptive (i.e. for users with enabled cookie warnings). If name is specified, the name of the current session is changed to its value.

**Comment by ChristophWurst** [comment]
We use that for the internal session wrapper already: https://github.com/owncloud/core/blob/57b999fde7cbfc06613c7a8bdff87febdb54b8fb/lib/private/Session/Internal.php#L45

**Comment by guruz** [comment]
Sorry, I guess it is https://secure.php.net/manual/de/function.session-id.php

**Comment by ChristophWurst** [comment]
Wow, didn't know that was possible :smile: 

Let me think about it, maybe this is indeed a solution to replace session tokens for devices :rocket:

**Pull request closed without merging**